// testplane/build/src/browser/cdp/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CDP_ERROR_CODE = exports.CDP_PING_MAX_SUBSEQUENT_FAILS = exports.CDP_PING_TIMEOUT = exports.CDP_PING_INTERVAL = exports.CDP_MAX_REQUEST_ID = exports.CDP_REQUEST_RETRY_BASE_DELAY = exports.CDP_REQUEST_RETRIES = exports.CDP_REQUEST_TIMEOUT = exports.CDP_CONNECTION_RETRY_BASE_DELAY = exports.CDP_CONNECTION_RETRIES = exports.CDP_CONNECTION_TIMEOUT = void 0;
exports.CDP_CONNECTION_TIMEOUT = 15000; // 15 sec
exports.CDP_CONNECTION_RETRIES = 3;
exports.CDP_CONNECTION_RETRY_BASE_DELAY = 500;
exports.CDP_REQUEST_TIMEOUT = 15000; // 15 sec
exports.CDP_REQUEST_RETRIES = 3;
exports.CDP_REQUEST_RETRY_BASE_DELAY = 500;
exports.CDP_MAX_REQUEST_ID = 2147483647; // INT32_MAX
exports.CDP_PING_INTERVAL = 15000; // 15 sec
exports.CDP_PING_TIMEOUT = 10000; // 10 sec
exports.CDP_PING_MAX_SUBSEQUENT_FAILS = 2;
exports.CDP_ERROR_CODE = {
    MALFORMED_RESPONSE: -32810, // Custom error code
    SEND_FAILED: -32820, // Custom error code
    TIMEOUT: -32830, // Custom error code
    CONNECTION_TERMINATED: -32840, // Custom error code
};
