// testplane/build/src/browser/commands/assert-view/errors/image-diff-error.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImageDiffError = void 0;
const base_state_error_1 = require("./base-state-error");
/**
 * @category Errors
 */
class ImageDiffError extends base_state_error_1.BaseStateError {
    static create({ stateName, currImg, refImg, diffOpts, diffAreas = {}, diffBuffer, differentPixels, diffRatio, }) {
        return new this({ stateName, currImg, refImg, diffOpts, diffAreas, diffBuffer, differentPixels, diffRatio });
    }
    static fromObject(data) {
        const { diffBounds, diffClusters, ...rest } = data;
        return new this({ ...rest, diffAreas: { diffBounds, diffClusters } });
    }
    constructor({ stateName, currImg, refImg, diffOpts, diffAreas: { diffBounds, diffClusters } = {}, diffBuffer, differentPixels, diffRatio, }) {
        super(stateName, currImg, refImg);
        this.message = `images are different for "${stateName}" state`;
        this.diffOpts = diffOpts;
        this.diffBounds = diffBounds;
        this.diffClusters = diffClusters;
        this.diffBuffer = diffBuffer;
        this.differentPixels = differentPixels;
        this.diffRatio = diffRatio;
    }
    saveDiffTo(diffPath) {
        return Promise.resolve().then(() => __importStar(require("../../../../image"))).then(m => m.Image.buildDiff({ diff: diffPath, ...this.diffOpts }));
    }
}
exports.ImageDiffError = ImageDiffError;
