// testplane/build/src/browser/cdp/domains/css.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CDPCss = void 0;
const emitter_1 = require("../emitter");
/** @link https://chromedevtools.github.io/devtools-protocol/tot/CSS/ */
class CDPCss extends emitter_1.CDPEventEmitter {
    constructor(connection) {
        super();
        this._connection = connection;
    }
    /** @param sessionId result of "Target.attachToTarget" */
    /** @link https://chromedevtools.github.io/devtools-protocol/tot/CSS/#method-enable */
    async enable(sessionId) {
        return this._connection.request("CSS.enable", { sessionId });
    }
    /** @param sessionId result of "Target.attachToTarget" */
    /** @link https://chromedevtools.github.io/devtools-protocol/tot/CSS/#method-disable */
    async disable(sessionId) {
        return this._connection.request("CSS.disable", { sessionId });
    }
    /** @link https://chromedevtools.github.io/devtools-protocol/tot/CSS/#method-getStyleSheetText */
    async getStyleSheetText(sessionId, styleSheetId) {
        return this._connection.request("CSS.getStyleSheetText", { sessionId, params: { styleSheetId } });
    }
    /** @link https://chromedevtools.github.io/devtools-protocol/tot/CSS/#method-startRuleUsageTracking */
    async startRuleUsageTracking(sessionId) {
        return this._connection.request("CSS.startRuleUsageTracking", { sessionId });
    }
    /** @link https://chromedevtools.github.io/devtools-protocol/tot/CSS/#method-stopRuleUsageTracking */
    async stopRuleUsageTracking(sessionId) {
        return this._connection.request("CSS.stopRuleUsageTracking", { sessionId });
    }
}
exports.CDPCss = CDPCss;
