// testplane/build/src/browser/cdp/domains/profiler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CDPProfiler = void 0;
const emitter_1 = require("../emitter");
/** @link https://chromedevtools.github.io/devtools-protocol/1-3/Profiler/ */
class CDPProfiler extends emitter_1.CDPEventEmitter {
    constructor(connection) {
        super();
        this._connection = connection;
    }
    /**
     * @param sessionId result of "Target.attachToTarget"
     * @link https://chromedevtools.github.io/devtools-protocol/1-3/Profiler/#method-disable
     */
    async disable(sessionId) {
        return this._connection.request("Profiler.disable", { sessionId });
    }
    /**
     * @param sessionId result of "Target.attachToTarget"
     * @link https://chromedevtools.github.io/devtools-protocol/1-3/Profiler/#method-enable
     */
    async enable(sessionId) {
        return this._connection.request("Profiler.enable", { sessionId });
    }
    /** @link https://chromedevtools.github.io/devtools-protocol/1-3/Profiler/#method-startPreciseCoverage */
    async startPreciseCoverage(sessionId, params) {
        return this._connection.request("Profiler.startPreciseCoverage", { sessionId, params });
    }
    /** @link https://chromedevtools.github.io/devtools-protocol/1-3/Profiler/#method-stopPreciseCoverage */
    async stopPreciseCoverage(sessionId) {
        return this._connection.request("Profiler.stopPreciseCoverage", { sessionId });
    }
    /** @link https://chromedevtools.github.io/devtools-protocol/1-3/Profiler/#method-takePreciseCoverage */
    async takePreciseCoverage(sessionId) {
        return this._connection.request("Profiler.takePreciseCoverage", { sessionId });
    }
}
exports.CDPProfiler = CDPProfiler;
