// @testplane/wdio-types/build/index.cjs
const __importMetaUrl = require('url').pathToFileURL(__filename).href;
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  MESSAGE_TYPES: () => MESSAGE_TYPES
});
module.exports = __toCommonJS(index_exports);

// src/Workers.ts
var MESSAGE_TYPES = /* @__PURE__ */ ((MESSAGE_TYPES2) => {
  MESSAGE_TYPES2[MESSAGE_TYPES2["consoleMessage"] = 0] = "consoleMessage";
  MESSAGE_TYPES2[MESSAGE_TYPES2["commandRequestMessage"] = 1] = "commandRequestMessage";
  MESSAGE_TYPES2[MESSAGE_TYPES2["commandResponseMessage"] = 2] = "commandResponseMessage";
  MESSAGE_TYPES2[MESSAGE_TYPES2["hookTriggerMessage"] = 3] = "hookTriggerMessage";
  MESSAGE_TYPES2[MESSAGE_TYPES2["hookResultMessage"] = 4] = "hookResultMessage";
  MESSAGE_TYPES2[MESSAGE_TYPES2["expectRequestMessage"] = 5] = "expectRequestMessage";
  MESSAGE_TYPES2[MESSAGE_TYPES2["expectResponseMessage"] = 6] = "expectResponseMessage";
  MESSAGE_TYPES2[MESSAGE_TYPES2["expectMatchersRequest"] = 7] = "expectMatchersRequest";
  MESSAGE_TYPES2[MESSAGE_TYPES2["expectMatchersResponse"] = 8] = "expectMatchersResponse";
  MESSAGE_TYPES2[MESSAGE_TYPES2["coverageMap"] = 9] = "coverageMap";
  MESSAGE_TYPES2[MESSAGE_TYPES2["customCommand"] = 10] = "customCommand";
  MESSAGE_TYPES2[MESSAGE_TYPES2["initiateBrowserStateRequest"] = 11] = "initiateBrowserStateRequest";
  MESSAGE_TYPES2[MESSAGE_TYPES2["initiateBrowserStateResponse"] = 12] = "initiateBrowserStateResponse";
  MESSAGE_TYPES2[MESSAGE_TYPES2["browserTestResult"] = 13] = "browserTestResult";
  return MESSAGE_TYPES2;
})(MESSAGE_TYPES || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MESSAGE_TYPES
});
